File: 30% Group Project -Assessment 3/AtanComelPretty/c/server/server_3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <arpa/inet.h>
#include <mysql/mysql.h>
#include <unistd.h>
#include <time.h>

void update_database(const char *username) {
    MYSQL *conn = mysql_init(NULL);
    if (conn == NULL) {
        fprintf(stderr, "mysql_init() failed\n");
        return;
    }

    if (mysql_real_connect(conn, "192.168.32.2", "root", "root", "testdb", 0, NULL, 0) == NULL) {
        fprintf(stderr, "mysql_real_connect() failed: %s\n", mysql_error(conn));
        mysql_close(conn);
        return;
    }

    char query[256];
    snprintf(query, sizeof(query), "UPDATE user_points SET points = points + 1, datetime_stamp = NOW() WHERE username = '%s';", username);

    if (mysql_query(conn, query)) {
        fprintf(stderr, "UPDATE failed: %s\n", mysql_error(conn));
    } else {
        printf("Database updated successfully for user 3: %s\n", username);
    }

    mysql_close(conn);
}

void get_points(const char *username, char *response) {
    MYSQL *conn = mysql_init(NULL);
    if (conn == NULL) {
        fprintf(stderr, "mysql_init() failed\n");
        strcpy(response, "Error: Unable to initialize MySQL");
        return;
    }

    if (mysql_real_connect(conn, "192.168.32.2", "root", "root", "testdb", 0, NULL, 0) == NULL) {
        fprintf(stderr, "mysql_real_connect() failed: %s\n", mysql_error(conn));
        strcpy(response, "Error: Unable to connect to MySQL");
        mysql_close(conn);
        return;
    }

    char query[256];
    snprintf(query, sizeof(query), "SELECT points FROM user_points WHERE username = '%s';", username);

    if (mysql_query(conn, query)) {
        fprintf(stderr, "SELECT failed: %s\n", mysql_error(conn));
        strcpy(response, "Error: Query execution failed");
    } else {
        MYSQL_RES *result = mysql_store_result(conn);
        if (result && mysql_num_rows(result) > 0) {
            MYSQL_ROW row = mysql_fetch_row(result);
            snprintf(response, 256, "User %s has %s points.", username, row[0]);
        } else {
            snprintf(response, 256, "User %s not found.", username);
        }
        mysql_free_result(result);
    }

    mysql_close(conn);
}

int main() {
    int sockfd;
    struct sockaddr_in server_addr, client_addr;
    char buffer[1024];
    char response[1024];
    socklen_t addr_size;
    time_t last_update_time = 0;

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(1035);  // Port matching Client 3 (c_client3)
    server_addr.sin_addr.s_addr = INADDR_ANY;

    bind(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr));
    printf("Server 3 running...\n");

    while (1) {
        // Check if it's time for a periodic update
        time_t current_time = time(NULL);
        if (current_time - last_update_time >= 30) {  // Every 5 seconds
            update_database("Atan_Comel3");
            last_update_time = current_time;
        }

        // Handle client requests
        addr_size = sizeof(client_addr);
        ssize_t recv_len = recvfrom(sockfd, buffer, sizeof(buffer) - 1, MSG_DONTWAIT, (struct sockaddr *)&client_addr, &addr_size);

        if (recv_len > 0) {
            buffer[recv_len] = '\0';  // Null-terminate the buffer
            printf("Received message from client 3: %s\n", buffer);

            if (strncmp(buffer, "GET_POINTS:", 11) == 0) {
                // Handle GET_POINTS request
                char *username = buffer + 11; // Extract the username
                get_points(username, response);
            } else if (strncmp(buffer, "UPDATE:", 7) == 0) {
                // Handle UPDATE request
                char *username = buffer + 7; // Extract the username
                update_database(username);
                strcpy(response, "Updated");
            } else {
                strcpy(response, "Error: Invalid request. Please send 'UPDATE:<username>' or 'GET_POINTS:<username>'.");
            }

            // Send response back to client
            sendto(sockfd, response, strlen(response), 0, (struct sockaddr *)&client_addr, addr_size);
        }

        // Small sleep to prevent CPU overuse
        usleep(100000);  // 100 ms
    }

    close(sockfd);
    return 0;
}

File: 30% Group Project -Assessment 3/AtanComelPretty/phyton/server/server_3.py
import socket
import mysql.connector
from mysql.connector import Error
import time
import threading

# Update points for the user in the database
def update_database(username):
    try:
        # Connect to MySQL
        conn = mysql.connector.connect(
            host="192.168.32.2",  # Use the correct host IP address for MySQL
            user="root",
            password="root",
            database="testdb"
        )
        if conn.is_connected():
            cursor = conn.cursor()
            query = f"""
                UPDATE user_points
                SET points = points + 1, datetime_stamp = NOW()
                WHERE username = %s;
            """
            cursor.execute(query, (username,))
            conn.commit()
            print(f"Database updated successfully for user: {username}")
        else:
            print("Connection to MySQL failed")
    except Error as e:
        print(f"Error: {e}")
    finally:
        if conn.is_connected():
            cursor.close()
            conn.close()

# Get points for the user from the database
def get_points(username):
    try:
        # Connect to MySQL
        conn = mysql.connector.connect(
            host="192.168.32.2",  # Use the correct host IP address for MySQL
            user="root",
            password="root",
            database="testdb"
        )
        if conn.is_connected():
            cursor = conn.cursor()
            query = f"SELECT points FROM user_points WHERE username = %s;"
            cursor.execute(query, (username,))
            result = cursor.fetchone()
            if result:
                return f"User {username} has {result[0]} points."
            else:
                return f"User {username} not found."
        else:
            return "Connection to MySQL failed"
    except Error as e:
        return f"Error: {e}"
    finally:
        if conn.is_connected():
            cursor.close()
            conn.close()

# Function to periodically update points every 3 seconds
def periodic_update(username):
    while True:
        time.sleep(30)  # Wait for 3 seconds
        print("Performing scheduled database update...")
        update_database(username)  # Update the database periodically for the same user

def main():
    # Set up the server
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    server_socket.bind(("0.0.0.0", 1033))  # Listen on port 1033 for server_3
    print("Python Server 3 running...")

    # Fixed username for Atan_Pretty3
    username = "Atan_Pretty3"  # Changed to Atan_Pretty3

    # Start the periodic update in a separate thread so it runs independently
    update_thread = threading.Thread(target=periodic_update, args=(username,))
    update_thread.daemon = True  # Ensures the thread will exit when the main program exits
    update_thread.start()

    while True:
        data, client_addr = server_socket.recvfrom(1024)
        message = data.decode().strip()
        print(f"Received message from client: {message}")

        # Handle requests based on message type
        if message.startswith("GET_POINTS:"):
            response = get_points(username)
        elif message.startswith("UPDATE:"):
            update_database(username)
            response = "Updated"
        else:
            response = "Error: Invalid request. Please send 'UPDATE:<username>' or 'GET_POINTS:<username>'."

        # Send the response back to the client
        server_socket.sendto(response.encode(), client_addr)

if __name__ == "__main__":
    main()
